TestRunner.psm1

```diff
@@ -81,84 +81,218 @@ function Invoke-RemoteCommand($server, [PSCredential] $credential, $argumentList
-
-        OPEN reorganizeOrRebuildCommands_cursor;
+function Write-ErrorLog {
+    param(
+        [Parameter(Mandatory=$True)]
+        [string]
+        $Message
+    )
+
+    Write-Log -Message $Message -Level "ERROR"
+}
+
+function Write-InfoLog {
+    param(
+        [Parameter(Mandatory=$True)]
+        [string]
+        $Message
+    )
+
+    Write-Log -Message $Message -Level "INFO"
+}
+
+function Write-DebugLog {
+    param(
+        [Parameter(Mandatory=$True)]
+        [string]
+        $Message
+    )
+
+    if ("DEBUG" -eq $env:PERF_LOG_LEVEL) {
+        Write-Log -Message $Message -Level "DEBUG"
+    }
+}
+
+function Reset-OdsDatabase {
+    param(
+        [Parameter(Mandatory=$True)]
+        [string]
+        $BackupFilename,
+
+        [Parameter(Mandatory=$True)]
+        [string]
+        $DatabaseName
+    )
+    # This import needs to be here, not at the top, so that this function can be
+    # executed on a remote server.
+    Import-Module SqlServer
+
+    $command = @"
+DECLARE @statement NVARCHAR(1000) = N'ALTER DATABASE ' + QUOTENAME(N'$DatabaseName') + N' SET SINGLE_USER';
+EXEC sys.sp_executesql @statement;
+"@
+
+    Write-DebugLog $command
+    Invoke-SqlCmd -Database "master" -Query $command -QueryTimeout 0
+
+    $command = @"
+DECLARE @BackupFilename NVARCHAR(128) = N'$BackupFilename';
+DECLARE @DatabaseName NVARCHAR(128) = N'$DatabaseName';
+DECLARE @statement NVARCHAR(1000);
+
+DECLARE @MdfPath NVARCHAR(256), @LdfPath NVARCHAR(256);
+SELECT @MdfPath = CAST(SERVERPROPERTY('InstanceDefaultDataPath') as NVARCHAR(256)),
+        @LdfPath = CAST(SERVERPROPERTY('InstanceDefaultLogPath') as NVARCHAR(256));
+
+-- Table definition below based on documentation at
+-- https://docs.microsoft.com/en-us/sql/t-sql/statements/restore-statements-filelistonly-transact-sql
+DECLARE @bak_files TABLE (
+    LogicalName nvarchar(128),
+    PhysicalName nvarchar(260),
+    [Type] char(1),
+    FileGroupName nvarchar(128),
+    Size numeric(20,0),
+    MaxSize numeric(20,0),
+    FileID bigint,
+    CreateLSN numeric(25,0),
+    DropLSN numeric(25,0),
+    UniqueID uniqueidentifier,
+    ReadOnlyLSN numeric(25,0),
+    ReadWriteLSN numeric(25,0),
+    BackupSizeInBytes bigint,
+    SourceBlockSize int,
+    FileGroupID int,
+    LogGroupGUID uniqueidentifier,
+    DifferentialBaseLSN numeric(25,0),
+    DifferentialBaseGUID uniqueidentifier,
+    IsReadOnly bit,
+    IsPresent bit,
+    TDEThumbprint varbinary(32),
+    SnapshotURL nvarchar(360)
+);
+
+SET @statement = N'RESTORE FILELISTONLY FROM DISK = @file';
+
+INSERT INTO @bak_files
+EXEC sys.sp_executesql @statement, N'@file NVARCHAR(128)', @file = @BackupFilename;
+
+SET @statement = 'RESTORE DATABASE @db FROM DISK = @disk WITH '
+SELECT @statement = @statement + 'MOVE N''' + LogicalName + ''' TO N''' +
+    CASE [Type] WHEN 'D' THEN @MdfPath + LogicalName +'.mdf'', '
+                WHEN 'L' THEN @LdfPath + LogicalName + '.ldf'', '
```

Comment:
Restores with the logical name from the backup. That may be ok for our testing use case here.

Reply:
There was some error that prompted me to do that, though I'm not remembering precisely what that was.